crates/storage/nippy-jar/src/cursor.rs

```diff
@@ -0,0 +1,215 @@
+    Filter, NippyJar, NippyJarError, PerfectHashingFunction, Row,
+};
+use memmap2::Mmap;
+use serde::{de::Deserialize, ser::Serialize};
+use std::{clone::Clone, fs::File};
+use sucds::int_vectors::Access;
+use zstd::bulk::Decompressor;
+
+/// Simple cursor implementation to retrieve data from [`NippyJar`].
+pub struct NippyJarCursor<'a, H> {
+    /// [`NippyJar`] which holds most of the required configuration to read from the file.
+    jar: &'a NippyJar<H>,
+    /// Optional dictionary decompressors.
+    zstd_decompressors: Option<Vec<Decompressor<'a>>>,
+    /// Data file.
+    #[allow(unused)]
+    file_handle: File,
+    /// Data file.
+    mmap_handle: Mmap,
+    /// Temporary buffer to unload data to (if necessary), without reallocating memory on each
+    /// retrieval.
+    tmp_buf: Vec<u8>,
+    /// Cursor row position.
+    row: u64,
+}
+
+impl<'a, H> std::fmt::Debug for NippyJarCursor<'a, H>
+where
+    H: Send + Sync + Serialize + for<'b> Deserialize<'b> + core::fmt::Debug,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("NippyJarCursor").field("config", &self.jar).finish_non_exhaustive()
+    }
+}
+
+impl<'a, H> NippyJarCursor<'a, H>
+where
+    H: Send + Sync + Serialize + for<'b> Deserialize<'b>,
+{
+    pub fn new(
+        config: &'a NippyJar<H>,
+        zstd_decompressors: Option<Vec<Decompressor<'a>>>,
+    ) -> Result<Self, NippyJarError> {
+        let file = File::open(config.data_path())?;
+
+        // SAFETY: File is read-only and its descriptor is kept alive as long as the mmap handle.
+        let mmap = unsafe { Mmap::map(&file)? };
+
+        Ok(NippyJarCursor {
+            jar: config,
+            zstd_decompressors,
+            file_handle: file,
+            mmap_handle: mmap,
+            tmp_buf: vec![],
+            row: 0,
+        })
+    }
+
+    /// Resets cursor to the beginning.
+    pub fn reset(&mut self) {
+        self.row = 0;
+    }
+
+    /// Returns a row, searching it by a key used during [`NippyJar::prepare_index`].
+    ///
+    /// **May return false positives.**
+    ///
+    /// Example usage would be querying a transactions file with a transaction hash which is **NOT**
+    /// stored in file.
+    pub fn row_by_key(&mut self, key: &[u8]) -> Result<Option<Row>, NippyJarError> {
+        if let (Some(filter), Some(phf)) = (&self.jar.filter, &self.jar.phf) {
+            // TODO: is it worth to parallize both?
+
+            // May have false positives
+            if filter.contains(key)? {
+                // May have false positives
+                if let Some(row_index) = phf.get_index(key)? {
+                    self.row =
+                        self.jar.offsets_index.access(row_index as usize).expect("built from same")
+                            as u64;
+                    return self.next_row()
+                }
+            }
+        }
+
+        Err(NippyJarError::UnsupportedFilterQuery)
```

Comment:
nit: it will return `NippyJarError::UnsupportedFilterQuery` if `filter.contains(key)` or `phf.get_index(key)` returned nothing, but instead it should return `Ok(None)`  I think?

Reply:
yes, you're right